pkg/resource/deploy/snapshot_test.go

```diff
@@ -77,3 +77,400 @@ func TestSnapshotWithUpdatedResources(t *testing.T) {
 	assert.NotSame(t, s, s1)
 	assert.Equal(t, s1.Resources[0].URN+"!", s.Resources[0].URN)
 }
+
+func TestSnapshotTopsort_PreservesValidSnapshots(t *testing.T) {
+	t.Parallel()
+
+	// Arrange.
+	cases := []struct {
+		name  string
+		given []*resource.State
+	}{
+		{
+			name:  "empty",
+			given: []*resource.State{},
+		},
+		{
+			name: "a single resource",
+			given: []*resource.State{
+				{URN: "urn:pulumi:stack::project::t::a"},
+			},
+		},
+		{
+			name: "two unrelated resources",
+			given: []*resource.State{
+				{URN: "urn:pulumi:stack::project::t::a"},
+				{URN: "urn:pulumi:stack::project::t::b"},
+			},
+		},
+		{
+			name: "two resources with a valid provider dependency",
+			given: []*resource.State{
+				{
```

Comment:
Shouldn't we pass these in _out of_ order? Otherwise the test would pass by just not doing anything

Reply:
That's the point right -- this test validates that a valid snapshot is left valid and not incorrectly mutated/broken. Happy to remove/rethink whether this is a useful test though if you have thoughts.

Reply:
Ah no, ok that makes sense